server/athenian/api/internal/miners/jira/issue.py

```diff
@@ -844,36 +832,94 @@ async def load(
-                    sql.and_(
-                        nprji.jira_acc == Issue.acc_id,
-                        nprji.jira_id == Issue.id,
-                    ),
+                    sql.and_(nprji.jira_acc == Issue.acc_id, nprji.jira_id == Issue.id),
                 ),
             )
-            .where(
-                sql.and_(
-                    node_id_cond,
-                    nprji.node_acc.in_(meta_ids),
-                ),
-            ),
+            .where(node_id_cond, nprji.node_acc.in_(meta_ids)),
+            mdb,
+            columns,
+            index=nprji.node_id.name,
         )
-        result = defaultdict(list)
-        for r in rows:
-            result[r[0]].append(r[1])
-        return result
+        res: dict[int, PullRequestJIRADetails] = {}
+        cls.append_from_df(res, df)
+        return res
+
+    @classmethod
+    @sentry_span
+    def append_from_df(
+        cls,
+        existing: dict[int, PullRequestJIRADetails],
+        df: pd.DataFrame,
+    ) -> None:
+        """Add the JIRA details in `df` to `existing` mapping from PR node IDs to JIRA."""
+        pr_node_ids = df.index.get_level_values(0).values
+        order = np.argsort(pr_node_ids)
+        unique_pr_ids, group_counts = np.unique(pr_node_ids[order], return_counts=True)
+        empty_cols = {}
+        payload_columns = []
+        for col in JIRAEntityToFetch.to_columns(JIRAEntityToFetch.EVERYTHING()):
+            df_name, dtype = cls._column_map[col.name]
+            if col.name not in df:
+                empty_cols[df_name] = np.array([], dtype=dtype)
+            else:
+                payload_columns.append(col.name)
+        pos = 0
+        for pr_id, group_count in zip(unique_pr_ids, group_counts):
+            indexes = order[pos : pos + group_count]
+            pos += group_count
+            # we can deduplicate. shall we? must benchmark the profit.
+            existing[pr_id] = PullRequestJIRADetails(
+                **{cls._column_map[c][0]: df[c].values[indexes] for c in payload_columns},
+                **empty_cols,
+            )
+
+    @classmethod
+    def apply_to_pr_facts(
+        self,
+        facts: PullRequestFactsMap,
+        mapping: dict[int, PullRequestJIRADetails],
+    ) -> None:
+        """Apply the jira mappings to the facts in PullRequestFactsMap, in place."""
+        empty = PullRequestJIRADetails.empty()
+        for (pr_id, _), pr_facts in facts.items():
+            try:
+                pr_facts.jira = mapping[pr_id]
+            except KeyError:
+                pr_facts.jira = empty
+
+    @classmethod
+    async def load_and_apply_to_pr_facts(
+        cls,
+        facts: PullRequestFactsMap,
+        entities: JIRAEntityToFetch | int,
+        meta_ids: tuple[int, ...],
+        mdb: DatabaseLike,
+    ) -> None:
+        """Load the jira mappings and apply it to the facts in PullRequestFactsMap, in place."""
+        pr_node_ids = [pr_node_id for pr_node_id, _ in facts]
```

Comment:
```suggestion
        pr_node_ids = np.fromiter((pr_node_id for pr_node_id, _ in facts), int, len(facts))
```
The size can be big, and we've got accelerated `IN (...)` generation for numpy arrays.